File: EngineKitGame/EngineKit/EngineKit/Items/Shape/Plane.m


#import "Plane.h"


@implementation Plane

+ (instancetype)plane {
    return [self new];
}

- (instancetype)init {
    if (self = [super initAndAddToScene]) {
        self.plane = [SCNPlane new];
    }
    return self;
}

- (instancetype)initWithWidth:(CGFloat)width height:(CGFloat)height {
    if (self = [super initAndAddToScene]) {
        self.plane = [SCNPlane planeWithWidth:width height:height];
    }
    return self;
}

- (Item *)deepCopy {
    Shape *newItem = [[self class] new];
    [self copyInfoTo:newItem];
    [super copyPhysicsTo:newItem];
    return newItem;
}

- (void)copyInfoTo:(Plane *)item {
    [super copyInfoTo:item];
    
    item.width = self.width;
    item.height = self.height;
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark - Property Overriding

- (void)setPlane:(SCNPlane *)plane {
    self.geometry = plane;
}

- (SCNPlane *)plane {
    return (SCNPlane *)self.geometry;
}


- (void)setWidth:(NSNumber *)width {
    [self assertTheresNoPhysicsBody];
    self.plane.width = width.doubleValue;
}

- (NSNumber *)width {
    return @(self.plane.width);
}

- (void)setHeight:(NSNumber *)height {
    [self assertTheresNoPhysicsBody];
    self.plane.height = height.doubleValue;
}

- (NSNumber *)height {
    return @(self.plane.height);
}

@end

File: Project/Project/Items/Shapes/Plane.swift

import Foundation
import SceneKit
import JavaScriptCore




@objc class Plane: Shape {
    
    var width: CGFloat {
        set { plane.width = newValue }
        get { return plane.width     }
    }
    
    var height: CGFloat {
        set { plane.height = newValue }
        get { return plane.height     }
    }
    
    override var size: CGFloat {
        set {
            width = newValue
            height = newValue
        }
        get {
            return max(width, height)
        }
    }
    
    var plane: SCNPlane {
        set { geometry = newValue        }
        get { return geometry as SCNPlane }
    }
    
    override init() {
        super.init(geometry: SCNPlane())
    }
}

